File: src/finder_pattern.cpp
#include "finder_pattern.h"

qr_cylinder::FinderPattern::FinderPattern()
{
    _up = new Arc();
    _down = new Arc();
    _left = new Arc();
    _right = new Arc();
}

void qr_cylinder::FinderPattern::SetPts(const vector<Point> &pts, int cols, int rows)
{
    _cols = cols;
    _rows = rows;
    if (!_orig_pts.empty())
        _orig_pts.clear();
    _orig_pts.assign(pts.begin(), pts.end());
    ResortPtsAndSetArc();
}

qr_cylinder::FinderPattern::~FinderPattern()
{
    delete _up;
    delete _down;
    delete _left;
    delete _right;
}

Point2f qr_cylinder::CalVecPara(const Point2f &vec_x, const Point2f &vec_y, const Point2f &vec_p)
{
    Point2f res;
    //Using solve function to get the answer
    Mat A, x, b;
    A.create(2, 2, CV_32FC1);
    A.at<float>(0, 0) = vec_x.x;
    A.at<float>(1, 0) = vec_x.y;
    A.at<float>(0, 1) = vec_y.x;
    A.at<float>(1, 1) = vec_y.y;
    b.create(2, 1, CV_32FC1);
    b.at<float>(0, 0) = vec_p.x;
    b.at<float>(1, 0) = vec_p.y;
    solve(A, b, x);
    res.x = x.at<float>(0, 0);
    res.y = x.at<float>(1, 0);
    return res;
}

void qr_cylinder::FinderPattern::Find4Corners(const vector<Point> &src_pts, vector<Point2f> &output_corners)
{
    if (!output_corners.empty())
        output_corners.clear();
    Rect rect = boundingRect(src_pts);
    Point base_pt = rect.tl();
    int width = rect.width + rect.width/2;
    int height = rect.height + rect.height/2;
    base_pt.x -= rect.width/4;
    base_pt.y -= rect.height/4;
    if (base_pt.x < 0)
        base_pt.x = 0;
    if (base_pt.y < 0)
        base_pt.y = 0;
    Mat tmp;
    tmp.create(height, width, CV_8UC1);
    tmp = Scalar::all(0);
    Point p1, p2;
    for (size_t i = 0; i < src_pts.size(); ++i) {
        p1 = src_pts[i];
        p2 = src_pts[(i+1)%src_pts.size()];
        p1.x -= base_pt.x;
        p1.y -= base_pt.y;
        p2.x -= base_pt.x;
        p2.y -= base_pt.y;
        line(tmp, p1, p2, Scalar(255, 255, 255));
    }
    double qualityLevel = 0.01;
    double minDistance = 8;
    int blockSize = 9;
    double k = 0.04;
    goodFeaturesToTrack(tmp, output_corners,
                        4, qualityLevel, minDistance,
                        Mat(), blockSize, false, k);
    for (size_t i = 0; i < output_corners.size(); ++i) {
        output_corners[i].x += base_pt.x;
        output_corners[i].y += base_pt.y;
    }
}

void qr_cylinder::FinderPattern::ResortPtsAndSetArc()
{
    Find4Corners(_orig_pts, _corners);

    // 找到四点的方位信息，UP-LEFT UP-RIGHT DOWN-LEFT DOWN-RIGHT
    // 并且找到他们的中心点CENTER, 以中心点到上的两个点作两个向量，
    // 计算各个点由它们表示的新坐标值的正负来判断他们是属于那一条边
    // UP DOWN LEFT RIGHT并且赋值给arc类
    // 以下这种方法并不是好的方法，是只考虑一个FP的情形所用到的方法
    // 所以需要找到一种方法使得
    // 上一层fp_manager确定方位型后重新来调整这个FP的上下左右
    sort(_corners.begin(), _corners.end(), LessY);
    vector<Point2f> up_pts, down_pts;
    up_pts.push_back(_corners[0]);
    up_pts.push_back(_corners[1]);
    down_pts.push_back(_corners[2]);
    down_pts.push_back(_corners[3]);
    sort(up_pts.begin(), up_pts.end(), LessX);
    sort(down_pts.begin(), down_pts.end(), LessX);
    if (!_corners.empty())
        _corners.clear();
    //按照up_left, up_right, down_right, down_left顺序存储角点
    _corners.push_back(up_pts[0]);
    _corners.push_back(up_pts[1]);
    _corners.push_back(down_pts[1]);
    _corners.push_back(down_pts[0]);

    Vec3d line1, line2, center_vec3d;
    line1[0] = _corners[0].y - _corners[2].y;
    line1[1] = _corners[2].x - _corners[0].x;
    line1[2] = _corners[0].x*_corners[2].y - _corners[0].y*_corners[2].x;
    line2[0] = _corners[1].y - _corners[3].y;
    line2[1] = _corners[3].x - _corners[1].x;
    line2[2] = _corners[1].x*_corners[3].y - _corners[1].y*_corners[3].x;
    center_vec3d[0] = line1[1]*line2[2] - line2[1]*line1[2];
    center_vec3d[1] = line2[0]*line1[2] - line1[0]*line2[2];
    center_vec3d[2] = line1[0]*line2[1] - line1[1]*line2[0];

    Point2f center;
    center.x = center_vec3d[0]/center_vec3d[2];
    center.y = center_vec3d[1]/center_vec3d[2];

    _center_point = center;

    Point2f vecX, vecY, vecP;
    vecX.x = up_pts[0].x - center.x;
    vecX.y = up_pts[0].y - center.y;
    vecY.x = up_pts[1].x - center.x;
    vecY.y = up_pts[1].y - center.y;
    vector<Point> UP, LEFT, RIGHT, DOWN;
    for (size_t i = 0; i < _orig_pts.size(); ++i) {
        vecP.x = _orig_pts[i].x - center.x;
        vecP.y = _orig_pts[i].y - center.y;
        Point2f vecRes = CalVecPara(vecX, vecY, vecP);
        if (vecRes.x > 0 && vecRes.y > 0)
            UP.push_back(_orig_pts[i]);
        else if (vecRes.x > 0 && vecRes.y < 0)
            LEFT.push_back(_orig_pts[i]);
        else if (vecRes.x < 0 && vecRes.y > 0)
            RIGHT.push_back(_orig_pts[i]);
        else if (vecRes.x < 0 && vecRes.y < 0)
            DOWN.push_back(_orig_pts[i]);
    }
    /*
     ** arc_num: 0~3
     ** 0: up arc
     ** 1: down arc
     ** 2: left arc
     ** 3: right arc
    */
    _up->SetPts(UP, true, true);
    _arcs[0] = _up;
    _down->SetPts(DOWN, true, true);
    _arcs[1] = _down;
    _left->SetPts(LEFT, true, false);
    _arcs[2] = _left;
    _right->SetPts(RIGHT, true, false);
    _arcs[3] = _right;
}

Point qr_cylinder::FinderPattern::GetCenterPoint()
{
    return _center_point;
}

void qr_cylinder::FinderPattern::DrawFp(Mat &draw_pic)
{
    _up->DrawArc(draw_pic, Scalar(255, 0, 0));
    _down->DrawArc(draw_pic, Scalar(0, 255, 0));
    _left->DrawArc(draw_pic, Scalar(0, 255, 255));
    _right->DrawArc(draw_pic, Scalar(255, 255, 0));
}

/*
 ** arc_num: 0~3
 ** 0: up arc
 ** 1: down arc
 ** 2: left arc
 ** 3: right arc
*/
void qr_cylinder::FinderPattern::PushArcPtsTo(unsigned int arc_num, vector<Point> &dst)
{
    _arcs[arc_num%4]->CopyPtsTo(dst);
}

File: android/app/src/main/java/com/pangomicro/zpfeng/cylinder_qrcode/FinderPattern.java
package com.pangomicro.zpfeng.cylinder_qrcode;

import org.opencv.core.CvType;
import org.opencv.core.Mat;
import org.opencv.core.MatOfPoint;
import org.opencv.core.Point;
import org.opencv.core.Rect;
import org.opencv.core.Scalar;
import org.opencv.imgproc.Imgproc;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import Jama.Matrix;

/**
 * Created by zczx1 on 2016/4/23.
 */
public class FinderPattern {
    public FinderPattern() {
        up = new Arc();
        down = new Arc();
        left = new Arc();
        right = new Arc();
        center_pts = new Point(-1, -1);
    }

    public boolean setFps(List<Point> pts) {
        if (pts.size() < 8)
            return false;
        corners = new ArrayList<>();
        orig_pts = new ArrayList<>();
        orig_pts.addAll(pts);
        find4Corners(orig_pts, corners);
        if (corners.size() != 4)
            return false;
        return resortPtsAndSetArc();
    }

    public void drawFp(Mat drawPic) {
        Imgproc.circle(drawPic, corners.get(0), 3, new Scalar(255, 0, 0), -1);
        Imgproc.circle(drawPic, corners.get(1), 3, new Scalar(0, 255, 0), -1);
        Imgproc.circle(drawPic, corners.get(2), 3, new Scalar(0, 0, 255), -1);
        Imgproc.circle(drawPic, corners.get(3), 3, new Scalar(255, 255, 0), -1);

        up.drawArc(drawPic, new Scalar(255, 0, 0));
        down.drawArc(drawPic, new Scalar(0, 255, 0));
        left.drawArc(drawPic, new Scalar(0, 0, 255));
        right.drawArc(drawPic, new Scalar(255, 255, 0));
    }

    public void pushArcPtsTo(int arc_num, List<Point> dst) {
        int i = arc_num%4;
        switch (i) {
            case 0:
                up.copyPtsTo(dst);
                break;
            case 1:
                down.copyPtsTo(dst);
                break;
            case 2:
                left.copyPtsTo(dst);
                break;
            case 3:
                right.copyPtsTo(dst);
                break;
        }
    }

    public void swapArcs() {
        Arc t1, t2;
        t1 = up;
        t2 = down;
        up = left;
        down = right;
        left = t1;
        right = t2;
        if (up.getCenter_point().y > down.getCenter_point().y) {
            t1 = up;
            up = down;
            down = t1;
        }
        if (left.getCenter_point().x > right.getCenter_point().x) {
            t1 = left;
            left = right;
            right = t1;
        }
    }

    public boolean isArcsTypeCorrect(Point vecP, boolean isVerticalX) {
        double res1, res2;
        res1 = calAbsCosValue(vecP, vecUp);
        res1 += calAbsCosValue(vecP, vecDown);
        res2 = calAbsCosValue(vecP, vecLeft);
        res2 += calAbsCosValue(vecP, vecRight);
        if (isVerticalX) {
            if (res2 <= res1)
                return false;
        } else {
            if (res2 >= res1)
                return false;
        }
        return true;
    }

    private double calAbsCosValue(Point vec1, Point vec2) {
        double res;
        res = (vec1.x*vec2.x + vec1.y*vec2.y)/(
                Math.sqrt(vec1.x*vec1.x + vec1.y*vec1.y)*Math.sqrt(vec2.x*vec2.x + vec2.y*vec2.y));
        return Math.abs(res);
    }

    public Point getCenterPoint() {
        return center_pts;
    }

    private boolean resortPtsAndSetArc() {
        Collections.sort(corners, new sortLessY());
        List<Point> upPts, downPts;
        upPts = new ArrayList<>();
        downPts = new ArrayList<>();
        upPts.add(corners.get(0));
        upPts.add(corners.get(1));
        downPts.add(corners.get(2));
        downPts.add(corners.get(3));
        Collections.sort(upPts, new sortLessX());
        Collections.sort(downPts, new sortLessX());
        corners.clear();
        //按照up_left, up_right, down_right, down_left顺序存储角点
        corners.add(upPts.get(0));
        corners.add(upPts.get(1));
        corners.add(downPts.get(1));
        corners.add(downPts.get(0));

        vecUp = new Point(-1, -1);
        vecDown = new Point(-1, -1);
        vecLeft = new Point(-1, -1);
        vecRight = new Point(-1, -1);
        vecUp.x = corners.get(1).x - corners.get(0).x;
        vecUp.y = corners.get(1).y - corners.get(0).y;
        vecDown.x = corners.get(2).x - corners.get(3).x;
        vecDown.y = corners.get(2).y - corners.get(3).y;
        vecLeft.x = corners.get(3).x - corners.get(0).x;
        vecLeft.y = corners.get(3).y - corners.get(0).y;
        vecRight.x = corners.get(1).x - corners.get(2).x;
        vecRight.y = corners.get(1).y - corners.get(2).y;

        Vec3d line1, line2, center_vec3d;
        double a, b, c;
        a = corners.get(0).y - corners.get(2).y;
        b = corners.get(2).x - corners.get(0).x;
        c = corners.get(0).x*corners.get(2).y - corners.get(0).y*corners.get(2).x;
        line1 = new Vec3d(a, b, c);
        a = corners.get(1).y - corners.get(3).y;
        b = corners.get(3).x - corners.get(1).x;
        c = corners.get(1).x*corners.get(3).y - corners.get(1).y*corners.get(3).x;
        line2 = new Vec3d(a, b, c);
        a = line1.getB()*line2.getC() - line2.getB()*line1.getC();
        b = line2.getA()*line1.getC() - line1.getA()*line2.getC();
        c = line1.getA()*line2.getB() - line2.getA()*line1.getB();
        center_vec3d = new Vec3d(a, b, c);
        Point center = new Point(-1, -1);
        center.x = center_vec3d.getA()/center_vec3d.getC();
        center.y = center_vec3d.getB()/center_vec3d.getC();
        center_pts.x = center.x;
        center_pts.y = center.y;
        Point vecX, vecY, vecP;
        vecX = new Point(-1, -1);
        vecY = new Point(-1, -1);
        vecP = new Point(-1, -1);
        vecX.x = upPts.get(0).x - center.x;
        vecX.y = upPts.get(0).y - center.y;
        vecY.x = upPts.get(1).x - center.x;
        vecY.y = upPts.get(1).y - center.y;
        List<Point> UP, DOWN, LEFT, RIGHT;
        UP = new ArrayList<>();
        DOWN = new ArrayList<>();
        LEFT = new ArrayList<>();
        RIGHT = new ArrayList<>();
        for (int i = 0; i < orig_pts.size(); ++i) {
            vecP.x = orig_pts.get(i).x - center.x;
            vecP.y = orig_pts.get(i).y - center.y;
            Point vecRes = calVecPara(vecX, vecY, vecP);
            if (vecRes.x > 0 && vecRes.y > 0)
                UP.add(orig_pts.get(i));
            else if (vecRes.x > 0 && vecRes.y < 0)
                LEFT.add(orig_pts.get(i));
            else if (vecRes.x < 0 && vecRes.y > 0)
                RIGHT.add(orig_pts.get(i));
            else if (vecRes.x < 0 && vecRes.y < 0)
                DOWN.add(orig_pts.get(i));
        }

        if (!up.setPts(UP, true, true))
            return false;
        if (!down.setPts(DOWN, true, true))
            return false;
        if (!left.setPts(LEFT, true, false))
            return false;
        if (!right.setPts(RIGHT, true, false))
            return false;
        return true;
    }

    private void find4Corners(List<Point> src_pts, List<Point> dst_pts) {
        MatOfPoint mop;
        mop = new MatOfPoint();
        mop.fromList(src_pts);
        Rect rect = Imgproc.boundingRect(mop);
        Point basePt = rect.tl();
        int width = rect.width + rect.width/2;
        int height = rect.height + rect.height/2;
        basePt.x -= rect.width/4;
        basePt.y -= rect.height/4;
        if (basePt.x < 0)
            basePt.x = 0;
        if (basePt.y < 0)
            basePt.y = 0;
        Mat tmp = new Mat(height, width, CvType.CV_8UC1, new Scalar(0));
        Point p1, p2;
        for (int i = 0; i < src_pts.size(); ++i) {
            p1 = src_pts.get(i).clone();
            p2 = src_pts.get((i+1)%src_pts.size()).clone();
            p1.x -= basePt.x;
            p1.y -= basePt.y;
            p2.x -= basePt.x;
            p2.y -= basePt.y;
            Imgproc.line(tmp, p1, p2, new Scalar(255));
        }
        double qualityLevel = 0.01;
        double minDistance = 8;
        int blockSize = 9;
        double k = 0.04;
        mop  = new MatOfPoint();
        Imgproc.goodFeaturesToTrack(tmp, mop,
                4, qualityLevel, minDistance,
                new Mat(), blockSize, false, k);
        Point[] pts = mop.toArray();
        for (int i = 0; i < 4; ++i) {
            Point tp = new Point(-1, -1);
            tp.x = pts[i].x;
            tp.y = pts[i].y;
            tp.x += basePt.x;
            tp.y += basePt.y;
            dst_pts.add(tp);
        }

    }

    private Point calVecPara(Point vecX, Point vecY, Point vecP) {
        Point res = new Point(-1, -1);
        Matrix A, x, b;
        A = new Matrix(2, 2);
        A.set(0, 0, vecX.x);
        A.set(1, 0, vecX.y);
        A.set(0, 1, vecY.x);
        A.set(1, 1, vecY.y);
        b = new Matrix(2, 1);
        b.set(0, 0, vecP.x);
        b.set(1, 0, vecP.y);
        x = A.solve(b);
        res.x = x.get(0, 0);
        res.y = x.get(1, 0);
        return res;
    }

    private Arc up, down, left, right;
    private List<Point> orig_pts;
    private List<Point> corners;
    private Point center_pts;
    Point vecUp, vecDown, vecLeft, vecRight;
}
